src/basic/fileio.c

```diff
@@ -856,35 +857,41 @@ int read_full_file_full(
         return read_full_stream_full(f, filename, offset, size, flags, ret_contents, ret_size);
 }
 
-int executable_is_script(const char *path, char **interpreter) {
-        _cleanup_free_ char *line = NULL;
-        size_t len;
-        char *ans;
+int script_get_interpreter(const char *path, char **ret) {
+        _cleanup_fclose_ FILE *f = NULL;
         int r;
 
         assert(path);
 
-        r = read_one_line_file(path, &line);
-        if (r == -ENOBUFS) /* First line overly long? if so, then it's not a script */
-                return 0;
-        if (r < 0)
-                return r;
+        f = fopen(path, "re");
+        if (!f)
+                return -errno;
 
-        if (!startswith(line, "#!"))
-                return 0;
+        char buf[STRLEN("#!")];
+        if (fread(buf, 1, sizeof(buf), f) != sizeof(buf)) {
+                if (feof(f))
+                        return -EINVAL;
```

Comment:
EINVAL? that's usually an error we return if we are incorrectly called. But if the data we process is invalid we should return something else.

I#d probably return -EBADMSG here or so.